File: camaleonte/camaleonte/python-cc/src/protocol/hash_protocol.py
from src.mediums.filesystem import HashEncoding
from .protocol import Protocol

from src.utils import get_hash_byte, set_hash_byte


class HashProtocol(Protocol):
    def __init__(self, filesystem: HashEncoding):
        self.filesystem = filesystem

    def encode_file(self, filepath: str, data: bytes) -> None:
        filedata = self.filesystem.read_content(filepath)
        # mine until desired hash
        filedata = set_hash_byte(filedata, ord(data))
        # update the file
        self.filesystem.write_content(filepath, filedata)

    def decode_file(self, filepath: str) -> bytes:
        filedata = self.filesystem.read_content(filepath)
        received_byte = get_hash_byte(filedata)
        # return as a 'bytes' type
        return chr(received_byte).encode()
    
    def data_per_file(self):
        return 1

File: camaleonte/camaleonte/metasploit/camaleonte/util/protocols/hash_protocol.rb
require_relative './protocol'
require_relative '../utils'

class HashProtocol
  include Protocol

  def initialize(filesystem)
    @filesystem = filesystem
  end

  # Modify filedata until first byte of hash is data
  def encode_file(filepath, data)
    filedata = @filesystem.read_content(filepath)
    # data is expected as bytes, get integer value of first byte
    desired_byte = data.bytes.first
    filedata = set_hash_byte(filedata, desired_byte)
    @filesystem.write_content(filepath, filedata)
  end

  # Decode byte from file data
  def decode_file(filepath)
    filedata = @filesystem.read_content(filepath)
    received_byte = get_hash_byte(filedata)
    # Return as single-byte string (Ruby String is byte sequence)
    received_byte.chr
  end

  # Only encoding/decoding one byte per file
  def data_per_file
    1
  end
end
